web/src/message_list_data_cache.ts

```diff
@@ -0,0 +1,33 @@
+import type {Filter} from "./filter";
+import type {MessageListData} from "./message_list_data";
+
+const cache: MessageListData[] = [];
+
+export function get(filter: Filter): MessageListData | undefined {
+    for (const cached_data of cache) {
+        if (cached_data.filter.equals(filter)) {
+            return cached_data;
+        }
+    }
+    return undefined;
+}
+
+export function add(message_list_data: MessageListData): void {
+    for (let i = 0; i < cache.length; i += 1) {
+        // Replace existing data in the cache if the filter is the same.
+        if (cache[i]?.filter.equals(message_list_data.filter)) {
+            cache[i] = message_list_data;
+            return;
+        }
+    }
+
+    cache.push(message_list_data);
+}
+
+export function all(): MessageListData[] {
+    return cache;
+}
```

Comment:
Should this be `return [...cache]` to avoid accidental mutation by callers?

Reply:
It could be that rather than `all_messages_data` being in the cache directly, this could be `return [...cache, all_messages_data]`, if we don't want to deal with `all_messages_data` potentially getting ejected. Not sure what's best.

Reply:
I think it's better for `all_messages_data` to be a part of the cache otherwise we will forget to include it in some loop we should have.

For making a copy, I think it's better for the caller to create a copy if it needs one.

Reply:
But the problem statement for not making a copy is that the caller mutates the list it has and ends up modifying the cache -- what I'm worried about is sloppy coding that calls this function being able to accidentally corrupt the cache.

Reply:
Yeah, updated to return a copy.